### ISSUE TITLE
tun/tap?

### ISSUE THREAD

**Issue by krzysztofantczak** [Microsoft/BashOnWindows#963]
Access to tun/tap devices so we could use openvpn/qemu inside of "bash" - that would be awesome!

**Comment by aseering** [comment]
@krzysztofantczak -- thanks for commenting!  This sounds related to #416 ; do the comments there help you out?  Not a fix, but at least a workaround.

**Comment by ericblade** [comment]
Also related to #915 ..

**Comment by rodrymbo** [comment]
For `openVPN`, I think starting it on Windows should usually result in WSL using its connection. 

If WSL is correctly using the Windows VPN (via passing connections off to Windows) the next issue is likely DNS resolution, since WSL will likely be set to use the original Windows DNS resolver, while the Windows VPN is likely to change that setting on the Windows side, and the original resolvers as set in `/etc/resolve.conf` are either blocked by the VPN or don't know about hosts on a private network on the VPN. In other words, the problem is not just enumeration or management of linux interfaces, but proper handling of DNS resolution.

As for `qemu` depending on tun/tap: it could probably get by just using the WSL network connection like everything else (that is, if it doesn't have to enumerate network interfaces or successfully complete a tun/tap installation or change the linux firewall or other things that aren't implemented yet).

Again, workarounds or descriptions of WSL shortcomings. The WSL problems definitely exist. But much of the interface difficulty is made worse because `/etc/resolv.conf` does not change in real time as resolver and/or routing settings on the Windows side change. Just implementing network interfaces on WSL is only part of the problem, unless WSL starts doing all of its own networking **independent** of Windows.

**Comment by krzysztofantczak** [comment]
@rodrymbo yes, i understand, in fact it works pretty well - windows qemu host is accessible inside WSL etc. however i'd like to keep my dev stack tight in place. I'm developing a system which uses local VM's, running ie. vbox/vagrant/qemu on windows and ie. my GO stack apps in WSL generates few awkward moments (i'm also manipulating VMs and their settings thru my stack). And i totally get it that i'll probably need to wait for it to work as we all would like to. Anyway this project looks really, really awesome as it is right now ;-)

**Comment by sunilmut** [comment]
If you have a VPN connection in Windows, there are currently issues with DNS resolution from bash. We are actively working to resolve that. But, at this time we do not have any plans of supporting VPN from bash. If that is important for you, please help us prioritize by casting your vote in our [user voice page](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/category/161892-bash). Also see #874

**Comment by krzysztofantczak** [comment]
Yeah, people seem to ignoring the fact, VPN was just an example. I'd like to just use the power of tun/tap devices, to that doesn't mean VPN only use case. So, it seems #874 should solve that for me. DNS issue seems not even minor here ;-)

**Comment by gengjiawen** [comment]
Anyone tried this on WSL 2 ?

**Comment by gengjiawen** [comment]
> Anyone tried this on WSL 2 ?

Works in WSL 2, just verified. Cheers 🍻
